### PR TITLE
[AMDGPU][GlobalIsel] Use isRegisterClassType for G_FREEZE and G_IMPLICIT_DEF

### PR DESCRIPTION
**Pull request by sstipano**
G_FREEZE was legal for <13 x S32> which caused an infinite loop in the combiner

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_implicit_def_legalizer.ll [hunk 1]
```diff
@@ -0,0 +1,120 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -global-isel -mtriple=amdgcn-mesa-mesa3d -mcpu=gfx1010 -verify-machineinstrs < %s | FileCheck -check-prefixes=GFX10 %s
+; RUN: llc -global-isel -mtriple=amdgcn-mesa-mesa3d -mcpu=gfx1100 -amdgpu-enable-delay-alu=0 -verify-machineinstrs < %s | FileCheck -check-prefixes=GFX11 %s
+
+target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-p7:160:256:256:32-p8:128:128-p9:192:256:256:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5-G1-ni:7:8:9-p32:32:32-p64:32:32-p65:32:32"
+target triple = "amdgcn--amdpal"
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_implicit_def_legalizer.ll:6]: Don't need these

File: llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_implicit_def_legalizer.ll [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+define amdgpu_cs void @_amdgpu_cs_main(i64 %0) {
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_implicit_def_legalizer.ll:8]: Use named values in tests

File: llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_implicit_def_legalizer.ll [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  %.ii125 = load float, ptr addrspace(1) %.i124, align 4
+  %.fr128.i1 = freeze float %.ii125
+  %44 = fadd float %.fr128.i1, 0.000000e+00
+  %45 = call float @llvm.fabs.f32(float %44)
+  %.i234 = getelementptr i8, ptr addrspace(1) %1, i64 48
+  %.ii235 = load float, ptr addrspace(1) %.i234, align 4
+  %.fr133.i2 = freeze float %.ii235
+  %46 = fadd float %.fr133.i2, 0.000000e+00
+  %47 = call float @llvm.fabs.f32(float %46)
+  br i1 %43, label %48, label %53
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_implicit_def_legalizer.ll:98]: Can probably reduce the test?

File: llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_implicit_def_legalizer.ll [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
+declare float @llvm.fabs.f32(float) #0
+
+; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
+declare noundef i64 @llvm.amdgcn.s.getpc() #0
+
+; uselistorder directives
+uselistorder ptr @llvm.fabs.f32, { 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0 }
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_implicit_def_legalizer.ll:118]: Definitely shouldn't need this

File: llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_implicit_def_legalizer.ll (hunk 3 above)
Comment (sstipanovic) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_implicit_def_legalizer.ll:98]: This is after running llvm-reduce

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_implicit_def_legalizer.ll:98]: You can usually manually get a bit further once you know what the issue is. This is big enough that I'm assuming we have a missing reduction technique somewhere

File: llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_implicit_def_legalizer.ll [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  %22 = getelementptr i8, ptr addrspace(1) %1, i64 28
+  %23 = load float, ptr addrspace(1) %22, align 4
+  %.fr128.i0 = freeze float %23
+  %24 = fadd float %.fr128.i0, 0.000000e+00
+  %25 = call float @llvm.fabs.f32(float %24)
+  %26 = getelementptr i8, ptr addrspace(1) %1, i64 36
+  %27 = load float, ptr addrspace(1) %26, align 4
+  %28 = call float @llvm.fabs.f32(float %27)
+  %29 = fcmp olt float %28, 1.000000e+00
+  %30 = and i1 %29, false
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_implicit_def_legalizer.ll:67]: Stuff like this should have folded out?

File: llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_implicit_def_legalizer.ll (hunk 3 above)
Comment (sstipanovic) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_implicit_def_legalizer.ll]: I've decided to use the simpler test case. For the previous test, it was the load/store vectorizer that produced <13 x float> and was a bit trickier to reduce further.

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 6]
```diff
@@ -889,10 +890,13 @@ AMDGPULegalizerInfo::AMDGPULegalizerInfo(const GCNSubtarget &ST_,
     .clampScalar(0, S16, S64);
 
   getActionDefinitionsBuilder({G_IMPLICIT_DEF, G_FREEZE})
-      .legalIf(isRegisterType(0))
+      // s1 and s16 are special cases because they have legal operations on
+      // them, but don't really occupy registers in the normal way.
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:894]: Duplicated comment

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 7]
```diff
@@ -889,10 +890,13 @@ AMDGPULegalizerInfo::AMDGPULegalizerInfo(const GCNSubtarget &ST_,
...
+      .legalIf(isRegisterClassType(0))
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:895]: Why do we have both "isRegisterType" and "isRegisterClassType"?

File: llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_legalizer.ll [hunk 8]
```diff
@@ -0,0 +1,87 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -global-isel -mtriple=amdgcn-mesa-mesa3d -mcpu=gfx1010 -verify-machineinstrs < %s | FileCheck -check-prefixes=GFX10 %s
+; RUN: llc -global-isel -mtriple=amdgcn-mesa-mesa3d -mcpu=gfx1100 -amdgpu-enable-delay-alu=0 -verify-machineinstrs < %s | FileCheck -check-prefixes=GFX11 %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_legalizer.ll:3]: Don't need -verify-machineinstrs

File: llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_legalizer.ll [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+define amdgpu_ps void @freeze13(ptr addrspace(1) %ptra, ptr addrspace(1) %ptrb) {
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/freeze_legalizer.ll:5]: Drop amdgpu_pses throughout

File: llvm/test/CodeGen/AMDGPU/GlobalISel/regbankselect.mir [hunk 10]
```diff
@@ -186,23 +184,6 @@ body: |
     %1:_(s32) = G_LOAD %0 :: (load (s32) from %ir.tmp1)
 ...
 
----
-name: non_power_of_2
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/regbankselect.mir:190]: Why was this deleted?

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp (hunk 7 above)
Comment (sstipanovic) [reply llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:895]: `isRegisterClassType` is checking against a predefined list of legal types. Currently only G_BITCAST uses it, the plan is to remove `isRegisterType` once every use is replaced. Introduced in https://github.com/llvm/llvm-project/pull/68189

File: llvm/test/CodeGen/AMDGPU/GlobalISel/regbankselect.mir (hunk 10 above)
Comment (sstipanovic) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/regbankselect.mir:190]: Because this test assumes that the input is already legalized, but S448 is not legal.

**Review by petar-avramovic — APPROVED** [review]

**Comment by chapuni** [comment]
Excuse me, I have reverted this. See,

https://lab.llvm.org/buildbot/#/builders/127/builds/891
https://lab.llvm.org/buildbot/#/builders/63/builds/1692
